File: templates/python/funnel_chart.py
"""funnel_chart: 漏斗图（流程转化率）."""
import sys, os; sys.path.insert(0, os.path.join(os.path.dirname(os.path.abspath(__file__)), '..', '..', '_utils', 'python'))
import numpy as np
import matplotlib.pyplot as plt
from theme import apply_theme
from palette import sequential

def make_figure(title='Funnel chart'):
    apply_theme(fig_size=(6.5, 5))
    stages = ['visit', 'sign-up', 'add to cart', 'checkout', 'paid']
    values = [10000, 4800, 2100, 1200, 850]
    cmap = sequential(hue='blue')
    fig, ax = plt.subplots()
    max_v = max(values)
    for i, (s, v) in enumerate(zip(stages, values)):
        w = v / max_v
        y = len(stages) - i - 1
        ax.barh(y, w,  height=0.8, color=cmap(0.3 + 0.6*v/max_v))
        ax.barh(y, -w, height=0.8, color=cmap(0.3 + 0.6*v/max_v))
        ax.text(0, y, f'{s}: {v}', ha='center', va='center', color='white', fontweight='bold')
    ax.set_yticks([]); ax.set_xticks([])
    ax.set_xlim(-1.1, 1.1); ax.set_title(title)
    ax.spines[['top', 'right', 'bottom', 'left']].set_visible(False)
    fig.tight_layout()
    return fig

if __name__ == '__main__':
    fig = make_figure(); fig.savefig(__file__.replace('.py', '.png'), dpi=150)

File: templates/matlab/funnel_chart.m
function fig = funnel_chart()
    addpath(fullfile(fileparts(mfilename('fullpath')),'..','..','_utils','matlab'));
    apply_theme();
    stages = {'visit','sign-up','add to cart','checkout','paid'};
    values = [10000 4800 2100 1200 850];
    cmap = palette('seq_blue');
    fig = figure('Position',[100 100 650 500]); hold on;
    max_v = max(values);
    for i = 1:numel(stages)
        w = values(i) / max_v; y = numel(stages) - i;
        c = cmap(round(80 + 175*values(i)/max_v), :);
        rectangle('Position',[-w, y, 2*w, 0.8], 'FaceColor', c, 'EdgeColor','w');
        text(0, y+0.4, sprintf('%s: %d', stages{i}, values(i)), ...
             'HorizontalAlignment','center', 'Color','w', 'FontWeight','bold');
    end
    xlim([-1.1 1.1]); ylim([-0.5 numel(stages)+0.2]);
    set(gca,'XTick',[],'YTick',[]); box off;
    title('Funnel chart');
end
